cpp/src/arrow/compute/kernels/scalar_temporal_benchmark.cc

```diff
@@ -0,0 +1,149 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <functional>
+
+#include "benchmark/benchmark.h"
+
+#include "arrow/compute/api_scalar.h"
+#include "arrow/compute/kernels/test_util.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/testing/random.h"
+#include "arrow/util/benchmark_util.h"
+
+namespace arrow {
+
+namespace compute {
+
+constexpr auto kSeed = 0x94378165;
+
+std::vector<int64_t> g_data_sizes = {kL2Size};
+static constexpr int64_t kInt64Min = -2000000000;  // 1906-08-16 20:26:40
+static constexpr int64_t kInt64Max = 2000000000;   // 2033-05-18 03:33:20
+
+void SetArgs(benchmark::internal::Benchmark* bench) {
+  for (const auto inverse_null_proportion : std::vector<ArgsType>({100, 0})) {
+    bench->Args({static_cast<ArgsType>(kL2Size), inverse_null_proportion});
+  }
+}
+
+using UnaryRoundingOp = Result<Datum>(const Datum&, const RoundTemporalOptions,
+                                      ExecContext*);
+using UnaryOp = Result<Datum>(const Datum&, ExecContext*);
+
+template <UnaryRoundingOp& Op, std::shared_ptr<DataType>& timestamp_type,
+          RoundTemporalOptions& options>
+static void BenchmarkTemporalRounding(benchmark::State& state) {
+  RegressionArgs args(state);
+
+  const int64_t array_size = args.size / sizeof(int64_t);
+
+  auto rand = random::RandomArrayGenerator(kSeed);
+  auto array = std::static_pointer_cast<NumericArray<Int64Type>>(
+      rand.Numeric<Int64Type>(array_size, kInt64Min, kInt64Max, args.null_proportion));
+  auto timestamp_array = std::make_shared<NumericArray<TimestampType>>(
+      timestamp_type, array->length(), array->data()->buffers[1],
+      array->data()->buffers[0], array->null_count());
```

Comment:
note I think you can do this with `Array::View` instead of having to do it manually

Reply:
Ah yes, much nicer.